### ISSUE TITLE
Unable to reset number of threads

### ISSUE THREAD

**Issue by ssomnath** [pydata/numexpr#364]
I'm not sure if I am doing things correct. I Think I am given the API. However, I am not able to reset the number of threads to a different value. I have tried this on my MacBook Pro (dual core) and on an IBM Power 9 machine.

```bash
In [1]: import numexpr as ne                                                    
n
In [2]: ne.nthreads                                                             
Out[2]: 4

In [3]: ne.set_num_threads(3)                                                   
Out[3]: 4

In [4]: ne.nthreads                                                             
Out[4]: 4

In [5]: ne.set_num_threads(1)                                                   
Out[5]: 3

In [6]: ne.nthreads                                                             
Out[6]: 4

In [7]: ne.__version__                                                          
Out[7]: '2.6.6'
```

Why is nthreads not changing?

**Comment by robbmcleod** [comment]
This is quite an older version, could you re-try with the latest release?

**Comment by ssomnath** [comment]
Tried out with the latest numexpr. Same behavior

```bash
In [1]: import numexpr as ne                                                    

In [2]: ne.__version__                                                          
Out[2]: '2.7.1'

In [3]: ne.nthreads                                                             
Out[3]: 4

In [4]: ne.set_num_threads(3)                                                   
Out[4]: 4

In [5]: ne.nthreads                                                             
Out[5]: 4

In [6]: ne.set_num_threads(1)                                                   
Out[6]: 3

In [7]: ne.nthreads                                                             
Out[7]: 4
```

**Comment by gdementen** [comment]
I just looked at the code and the behavior you describe is not really surprising as AFAICT nothing updates the value of the python-side module global (nthreads) except on first initialization. The actual number of threads used by the interpreter is stored in a C variable and its value is not mirrored in the Python module variable. In other words, the number of threads is actually set correctly, but you cannot access it via ne.nthreads, you can only know the value of it by changing it using set_num_threads.

As a workaround, you could use something like this (untested!!!) code:
```python
def get_nthreads():
   nthreads = ne.set_num_threads(1)
   ne.set_num_threads(nthreads)
   return nthreads
```

**Comment by robbmcleod** [comment]
Yeah I remember now. I'm pretty sure in `numexpr3` it's implemented properly so it should be pretty simple to add a `get_num_threads()` function to the C-API. Even if I didn't it's a trivial job.

**Comment by robbmcleod** [comment]
Anyone want to give the latest build a try? I added a `numexpr.get_num_threads()` function that directly interrogates the value in the virtual machine. I'll deprecate the return value for `numexpr.set_num_threads`, to be removed some time in the far future.

**Comment by ssomnath** [comment]
Thank you @robbmcleod  and @gdementen for clarifying. I need to check if numexpr3 is compatible with other packages I am using. If it is, I will look into swapping.

**Comment by robbmcleod** [comment]
I'm not suggesting you switch to `numexpr3` unless you need its extra capabilities. I made two pushes to `numexpr` 2.7.2dev, that should provide you with expected behaviour.

**Comment by ssomnath** [comment]
OK. Thanks!

**Comment by robbmcleod** [comment]
Is resolved in the 2.7.2 release. In the future we will stop returning the previous set value in `set_num_threads` as that's not really the behavior one would typically expected in a property setter.
